### PR TITLE
[5.1] Fix active url validation

### PR DESCRIPTION
**Pull request by cviebrock**
Fix for #11173

### CODE REVIEW

**Comment by c9n** [comment]
:+1:

File: src/Illuminate/Validation/Validator.php [hunk 1]
```diff
@@ -1235,7 +1235,9 @@ protected function validateUrl($attribute, $value)
      */
     protected function validateActiveUrl($attribute, $value)
     {
-        $url = str_replace(['http://', 'https://', 'ftp://'], '', strtolower($value));
+        if (! ($url = parse_url($value, PHP_URL_HOST))) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Validation/Validator.php]: Please assign the url variable first.

File: src/Illuminate/Validation/Validator.php [hunk 2]
```diff
@@ -1235,7 +1235,9 @@ protected function validateUrl($attribute, $value)
      */
     protected function validateActiveUrl($attribute, $value)
     {
-        $url = str_replace(['http://', 'https://', 'ftp://'], '', strtolower($value));
```
Comment (GrahamCampbell) [inline src/Illuminate/Validation/Validator.php]: Could you add a test case to show ftp urls still work please.

Comment (GrahamCampbell) [inline src/Illuminate/Validation/Validator.php]: Same for https.

**Comment by taylorotwell** [comment]
Could this be misleading maybe? What if the actual URL isn't active but the root domain is. Aren't we trying to validate the exact URL that is passed in?

**Comment by willrowe** [comment]
I believe if you were to pass `http://google.com/about` it would return `false` even if that page did exist since `checkdnsrr` can only validate host names and IP addresses not full URLs.

File: src/Illuminate/Validation/Validator.php (hunk 2 above)
Comment (cviebrock) [inline src/Illuminate/Validation/Validator.php]: The new logic simply ignores all protocols and only extracts the hostname from the given value.  So, even adding test cases with `ftp://` and `https://` protocols isn't going to really test anything, other than that PHP's internal `parse_url()` function works.

**Comment by cviebrock** [comment]
@taylorotwell: What @willrowe said is correct.  The current implementation fails when validating `https://google.com/about`.  In fact, it will fail on anything more than a protocol+base domain, since `checkdnsrr()` was meant for DNS records, not URLs.

I wasn't attempting to do a _full_ validation of the URL, which would require making an actual connection attempt.  That's a valid use-case, but I think would warrant a new validator method to prevent breaking issues (or, at least, to prevent extra network traffic when it's not expected).

File: src/Illuminate/Validation/Validator.php (hunk 1 above)
Comment (cviebrock) [inline src/Illuminate/Validation/Validator.php]: I've inverted the logic, which I think is clearer.

File: src/Illuminate/Validation/Validator.php [hunk 3]
```diff
@@ -1235,9 +1235,11 @@ protected function validateUrl($attribute, $value)
...
+        if ($url = parse_url($value, PHP_URL_HOST)) {
+            return checkdnsrr($url, 'A');
```
Comment (pascalschwientek) [inline src/Illuminate/Validation/Validator.php]: What about IPv6 and AAAA records?

Comment (cviebrock) [inline src/Illuminate/Validation/Validator.php]: Good point.  Those would've failed validation before as well.  Suggestion:

```php
return checkdnsrr($url, 'A') || checkdnsrr($url, 'AAAA');
```

Maybe?

Comment (lucasmichot) [inline src/Illuminate/Validation/Validator.php]: You can target relevant A and AAAA types using only one function call.
```php
return dns_get_record($url, DNS_A | DNS_AAAA) > 0;
```

Comment (cviebrock) [inline src/Illuminate/Validation/Validator.php]: :+1: implemented your first suggestion.

**Comment by GrahamCampbell** [comment]
Thanks for the contribution. I wonder if we should maybe just add this as a new validation rule rather than updating the existing one? Now that 5.1 is in it's LTS maintenance phase, new features will need to go to 5.2.

Ultimately though, it's up to Taylor to decide what he wants here.

**Comment by GrahamCampbell** [comment]
No need to retarget at 5.2 or anything for now is what I was getting at. ;)

**Comment by cviebrock** [comment]
Fair enough.  I still feel, even with my PR, that the `active_url` rule is "broken", in that it isn't really checking if the given URL is active.  I was simply trying to make it "less broken".  In that regard, I could argue for it making it into 5.1.  I'd also be in favour of a new rule for 5.2 that made an HTTP/FTP request to _really_ check a URL (although there are lots of reasons that might fail on valid URLs as well).

**Comment by taylorotwell** [comment]
If wanted to make a new rule that actually hits the given URL for a 200 response we could have something like `ping_url` but I think we should wait for community demand on that. This rule has been out for years and only just now is getting this fix which tells me it may not be the most highly used validation rule we have :smile:

**Comment by vogelor** [comment]
i know, i am late. but i find out that this will NOT work with IDN Domains like "müller.de"
you have to contert it to a ascii-domain.
you can use this function: http://php.net/manual/de/function.idn-to-ascii.php
or - for example - this class (because idn-to-ascii needs to be "installed")
https://github.com/true/php-punycode

**Pull request merged**